### PR TITLE
Final phase of source tree restructure

### PR DESCRIPTION
**Pull request by lstein**
# All python code has been moved under `invokeai`. All vestiges of `ldm` and `ldm.invoke` are now gone.

***You will need to run `pip install -e .` before the code will work again!***

Everything seems to be functional, but extensive testing is advised.

A guide to where the files have gone is forthcoming.

### CODE REVIEW

**Comment by psychedelicious** [comment]
1. is git history/blame retained through the reorganisation?
2. i have a large PR #2814 WIP for the frontend. i'll probably be finished by friday if not sooner. i would really like to get that into main and then have the reorganisation be after that. rebasing this PR on a different folder structure will be very tedious.

**Comment by Kyle0654** [comment]
> 1. is git history/blame retained through the reorganisation?
> 2. i have a large PR [[ui]: migrate all styling to chakra-ui theme #2814](https://github.com/invoke-ai/InvokeAI/pull/2814) WIP for the frontend. i'll probably be finished by friday if not sooner. i would really like to get that into main and then have the reorganisation be after that. rebasing this PR on a different folder structure will be very tedious.

Sounds like frontend files may not be moving?

**Comment by psychedelicious** [comment]
> Sounds like frontend files may not be moving?

oh, you're right. all good

**Comment by lstein** [comment]
> * is git history/blame retained through the reorganisation?
> * i have a large PR [[ui]: migrate all styling to chakra-ui theme #2814](https://github.com/invoke-ai/InvokeAI/pull/2814) WIP for the frontend. i'll probably be finished by friday if not sooner. i would really like to get that into main and then have the reorganisation be after that. rebasing this PR on a different folder structure will be very tedious.

The identity of the files remains intact through organization and `git blame` stays the same. If a file has been moved you can still often, but not always, merge onto it without conflicts. That being said, the frontend and backend/webui files have not moved at all, and aren't going to. Everything else is moving to join them!

**Comment by Kyle0654** [comment]
Looks like some `__pycache__` snuck in.

**Comment by Kyle0654** [comment]
It feels like the Stable Diffusion code should probably go in its own module under the backend folder, to keep it separate from app code (e.g. CLI or API). I noticed this most in the `/invokeai/backend/models` folder, where `models` here is referring to diffusion models, but `models` in an API usually refers to the data models used for serializing.

Then you could have `/invokeai/backend/app` with `~/cli`, `~/api`, and `~/core` (or whatever makes sense).

Two popular templates are roughly organized this way:
https://github.com/Buuntu/fastapi-react
https://github.com/arthurhenrique/cookiecutter-fastapi

**Comment by lstein** [comment]
> Looks like some `__pycache__` snuck in.

Removed

**Comment by lstein** [comment]
> It feels like the Stable Diffusion code should probably go in its own module under the backend folder, to keep it separate from app code (e.g. CLI or API). I noticed this most in the `/invokeai/backend/models` folder, where `models` here is referring to diffusion models, but `models` in an API usually refers to the data models used for serializing.
> 
> Then you could have `/invokeai/backend/app` with `~/cli`, `~/api`, and `~/core` (or whatever makes sense).
> 
> Two popular templates are roughly organized this way: https://github.com/Buuntu/fastapi-react https://github.com/arthurhenrique/cookiecutter-fastapi

How about `invokeai/backend/diffusion_models` ?

**Comment by Kyle0654** [comment]
> How about `invokeai/backend/ldm/models` ? I don't want to confuse people too much by renaming things too radically. Then the modules would move to `invokeai/backend/ldm/modules` and the utilities to `invokeai/backend/ldm/util`.

Things are moving around a bunch already. No better time to change names. Might be a good idea to separate from the `ldm` name too? Didn't diffusers supersede that?

**Comment by lstein** [comment]
> > How about `invokeai/backend/ldm/models` ? I don't want to confuse people too much by renaming things too radically. Then the modules would move to `invokeai/backend/ldm/modules` and the utilities to `invokeai/backend/ldm/util`.
> 
> Things are moving around a bunch already. No better time to change names. Might be a good idea to separate from the `ldm` name too? Didn't diffusers supersede that?

Diffusers are implementing latent diffusion models,  so I think we can legitimately keep the `ldm` name. Structure is now `invokeai/backend/ldm/{models, modules}`

File: .gitignore [hunk 1]
```diff
@@ -214,9 +214,9 @@ gfpgan/
 configs/models.yaml
 
 # weights (will be created by installer)
-models/ldm/stable-diffusion-v1/*.ckpt
-models/clipseg
-models/gfpgan
+# models/ldm/stable-diffusion-v1/*.ckpt
```
Comment (Kyle0654) [inline .gitignore:217]: Should these just be removed instead of commented out?

File: invokeai/backend/ldm/models/model_manager.py [hunk 2]
```diff
@@ -370,125 +353,6 @@ def _load_model(self, model_name: str):
             )
         return model, width, height, model_hash
 
-    def _load_ckpt_model(self, model_name, mconfig):
```
Comment (Kyle0654) [inline invokeai/backend/ldm/models/model_manager.py:373]: Do we still support ckpt models and just load them through a different path now?

**Review by Kyle0654 — APPROVED** [review]

**Comment by damian0815** [comment]
as i said in discord, just gonna hit "approve" on this

**Review by damian0815 — APPROVED** [review]

File: invokeai/backend/ldm/models/model_manager.py (hunk 2 above)
Comment (lstein) [reply invokeai/backend/ldm/models/model_manager.py:373]: We load .ckpt and .safetensors models by converting them into `diffusers` in RAM. This works 99% of the time.

File: .gitignore (hunk 1 above)
Comment (lstein) [reply .gitignore:217]: Good point.

**Review by damian0815 — APPROVED** [review]
👍

**Review by ebr — APPROVED** [review]
new structure makes a lot of sense :100: :+1:

**Comment by Kyle0654** [comment]
Could you undo the auto code formatting on at least the invocations (and add whatever needs to be added to avoid it)? It changed this:
```py
    prompt: Optional[str]     = Field(description="The prompt to generate an image from")
    seed: int                 = Field(default=-1, ge=-1, le=np.iinfo(np.uint32).max, description="The seed to use (-1 for a random seed)")
    steps: int                = Field(default=10, gt=0, description="The number of steps to use to generate the image")
    width: int                = Field(default=512, multiple_of=64, gt=0, description="The width of the resulting image")
    height: int               = Field(default=512, multiple_of=64, gt=0, description="The height of the resulting image")
    cfg_scale: float          = Field(default=7.5, gt=0, description="The Classifier-Free Guidance, higher values may result in a result closer to the prompt")
    sampler_name: SAMPLER_NAME_VALUES = Field(default="k_lms", description="The sampler to use")
    seamless: bool            = Field(default=False, description="Whether or not to generate an image that can tile without seams")
    model: str                = Field(default='', description="The model to use (currently ignored)")
    progress_images: bool     = Field(default=False, description="Whether or not to produce progress images during generation")
```
Into this:
```py
    prompt: Optional[str] = Field(description="The prompt to generate an image from")
    seed: int = Field(
        default=-1,
        ge=-1,
        le=np.iinfo(np.uint32).max,
        description="The seed to use (-1 for a random seed)",
    )
    steps: int = Field(
        default=10, gt=0, description="The number of steps to use to generate the image"
    )
    width: int = Field(
        default=512,
        multiple_of=64,
        gt=0,
        description="The width of the resulting image",
    )
    height: int = Field(
        default=512,
        multiple_of=64,
        gt=0,
        description="The height of the resulting image",
    )
    cfg_scale: float = Field(
        default=7.5,
        gt=0,
        description="The Classifier-Free Guidance, higher values may result in a result closer to the prompt",
    )
    sampler_name: SAMPLER_NAME_VALUES = Field(
        default="k_lms", description="The sampler to use"
    )
    seamless: bool = Field(
        default=False,
        description="Whether or not to generate an image that can tile without seams",
    )
    model: str = Field(default="", description="The model to use (currently ignored)")
    progress_images: bool = Field(
        default=False,
        description="Whether or not to produce progress images during generation",
    )
```
IMO I prefer the original formatting. Especially since the latter isn't even consistent (e.g. `steps` has been consolidated to a single line, whereas `cfg_scale` splits each parameter to a separate line).

**Comment by lstein** [comment]
> It's also a nitpick, but a lot of what's in `app` would typically go in `backend` in a client-server application (with the client living in `frontend`). We have a CLI as well (as opposed to the traditional API and web client), which complicates things some, but I do feel like having all the Python code in the same folder (to the extent that you could potentially even split off `backend` into its own repo) would make sense.

My thinking on this was to do the `generate` refactor (which means reimplementing nodes to run off of `generator` classes directly) and to integrate the nodes into the `frontend`/`backend` directories as the migration is finished.

**Comment by lstein** [comment]
> 

Done. Note that the way to disable autoformatting is like this:

```
#fmt: off
code code code
#fmt: on

**Comment by Kyle0654** [comment]
Do you mind hitting the rest of the invocations as well? E.g. the Img2Img one.

**Comment by lstein** [comment]
> Do you mind hitting the rest of the invocations as well? E.g. the Img2Img one.
> 
> Ack looks like I was too late x.x

Oh shoot. I thought all the invocations were at the bottom of the files. Will make a PR.

**Comment by mickr777** [comment]
@lstein fresh install on main when creating a image
```
Traceback (most recent call last):
  File "/home/invokeuser/InvokeAI/invokeai/backend/web/invoke_ai_web_server.py", line 1313, in generate_images
    self.generate.prompt2image(
  File "/home/invokeuser/InvokeAI/invokeai/backend/generate.py", line 539, in prompt2image
    generator = self.select_generator(
  File "/home/invokeuser/InvokeAI/invokeai/backend/generate.py", line 818, in select_generator
    return self._make_txt2img2img()
  File "/home/invokeuser/InvokeAI/invokeai/backend/generate.py", line 894, in _make_txt2img2img
    return self._load_generator(".txt2img2img", "Txt2Img2Img")
  File "/home/invokeuser/InvokeAI/invokeai/backend/generate.py", line 902, in _load_generator
    module = importlib.import_module(mn)
  File "/usr/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/invokeuser/InvokeAI/invokeai/backend/generator/txt2img2img.py", line 11, in <module>
    from ..models import PostprocessingSettings
ModuleNotFoundError: No module named 'invokeai.backend.models'

```

**Pull request merged**